File: php/laravel/tests/Feature/UserControllerTest.php
<?php

namespace Tests\Feature;

use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Foundation\Testing\WithFaker;
use Tests\TestCase;

class UserControllerTest extends TestCase
{
    /**
     * Test health check endpoint.
     */
    public function test_health_check(): void
    {
        $response = $this->get('/health');

        $response->assertStatus(200)
                ->assertJsonStructure([
                    'status',
                    'timestamp',
                    'service'
                ]);
    }

    /**
     * Test get all users.
     */
    public function test_get_all_users(): void
    {
        $response = $this->get('/users');

        $response->assertStatus(200)
                ->assertJsonStructure([
                    'users' => [
                        '*' => [
                            'id',
                            'name',
                            'email'
                        ]
                    ],
                    'total'
                ]);
    }

    /**
     * Test get user by ID.
     */
    public function test_get_user_by_id(): void
    {
        $response = $this->get('/users/1');

        $response->assertStatus(200)
                ->assertJsonStructure([
                    'id',
                    'name',
                    'email'
                ]);
    }

    /**
     * Test get non-existent user.
     */
    public function test_get_nonexistent_user(): void
    {
        $response = $this->get('/users/999');

        $response->assertStatus(404)
                ->assertJson([
                    'error' => 'User not found'
                ]);
    }

    /**
     * Test create user.
     */
    public function test_create_user(): void
    {
        $userData = [
            'name' => 'Test User',
            'email' => 'test@example.com'
        ];

        $response = $this->post('/users', $userData);

        $response->assertStatus(201)
                ->assertJsonStructure([
                    'id',
                    'name',
                    'email'
                ])
                ->assertJson([
                    'name' => 'Test User',
                    'email' => 'test@example.com'
                ]);
    }

    /**
     * Test create user with invalid data.
     */
    public function test_create_user_with_invalid_data(): void
    {
        $userData = [
            'name' => '',
            'email' => 'invalid-email'
        ];

        $response = $this->post('/users', $userData);

        $response->assertStatus(400)
                ->assertJsonStructure([
                    'error'
                ]);
    }

    /**
     * Test update user.
     */
    public function test_update_user(): void
    {
        $updateData = [
            'name' => 'Updated Name'
        ];

        $response = $this->put('/users/1', $updateData);

        $response->assertStatus(200)
                ->assertJsonStructure([
                    'id',
                    'name',
                    'email'
                ])
                ->assertJson([
                    'name' => 'Updated Name'
                ]);
    }

    /**
     * Test update non-existent user.
     */
    public function test_update_nonexistent_user(): void
    {
        $updateData = [
            'name' => 'Updated Name'
        ];

        $response = $this->put('/users/999', $updateData);

        $response->assertStatus(404)
                ->assertJson([
                    'error' => 'User not found'
                ]);
    }

    /**
     * Test delete user.
     */
    public function test_delete_user(): void
    {
        $response = $this->delete('/users/2');

        $response->assertStatus(200)
                ->assertJson([
                    'message' => 'User deleted successfully'
                ]);
    }

    /**
     * Test delete non-existent user.
     */
    public function test_delete_nonexistent_user(): void
    {
        $response = $this->delete('/users/999');

        $response->assertStatus(404)
                ->assertJson([
                    'error' => 'User not found'
                ]);
    }
}

File: java/springboot/src/test/java/com/example/apiserver/controller/UserControllerTest.java
package com.example.apiserver.controller;

import com.example.apiserver.dto.CreateUserRequest;
import com.example.apiserver.dto.UpdateUserRequest;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureTestMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureTestMvc
public class UserControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Test
    public void testHealthCheck() throws Exception {
        mockMvc.perform(get("/health"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status").value("healthy"))
                .andExpect(jsonPath("$.service").value("java-api-server"))
                .andExpect(jsonPath("$.timestamp").exists());
    }

    @Test
    public void testGetAllUsers() throws Exception {
        mockMvc.perform(get("/users"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.users").isArray())
                .andExpect(jsonPath("$.total").value(2));
    }

    @Test
    public void testGetUserById() throws Exception {
        mockMvc.perform(get("/users/1"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(1))
                .andExpect(jsonPath("$.name").value("John Doe"))
                .andExpect(jsonPath("$.email").value("john@example.com"));
    }

    @Test
    public void testGetUserByIdNotFound() throws Exception {
        mockMvc.perform(get("/users/999"))
                .andExpect(status().isNotFound())
                .andExpect(jsonPath("$.error").value("User not found"));
    }

    @Test
    public void testCreateUser() throws Exception {
        CreateUserRequest request = new CreateUserRequest("Test User", "test@example.com");
        
        mockMvc.perform(post("/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.name").value("Test User"))
                .andExpect(jsonPath("$.email").value("test@example.com"))
                .andExpect(jsonPath("$.id").exists());
    }

    @Test
    public void testCreateUserInvalidEmail() throws Exception {
        CreateUserRequest request = new CreateUserRequest("Test User", "invalid-email");
        
        mockMvc.perform(post("/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isBadRequest());
    }

    @Test
    public void testUpdateUser() throws Exception {
        UpdateUserRequest request = new UpdateUserRequest("Updated Name", null);
        
        mockMvc.perform(put("/users/1")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.name").value("Updated Name"))
                .andExpect(jsonPath("$.email").value("john@example.com"));
    }

    @Test
    public void testUpdateUserNotFound() throws Exception {
        UpdateUserRequest request = new UpdateUserRequest("Updated Name", null);
        
        mockMvc.perform(put("/users/999")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isNotFound());
    }

    @Test
    public void testDeleteUser() throws Exception {
        mockMvc.perform(delete("/users/2"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.message").value("User deleted successfully"));
    }

    @Test
    public void testDeleteUserNotFound() throws Exception {
        mockMvc.perform(delete("/users/999"))
                .andExpect(status().isNotFound())
                .andExpect(jsonPath("$.error").value("User not found"));
    }
}